### PR TITLE
feat: wire in the KS config needed for point queries (MINOR)

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Point queries need to be able to determine with KSQL node stores the data for a specific key in a state store. Kafka Streams supports this via its interactive queries functionality. For this to work, the KS `application.server` config must be set.

This change sees the `application.server` being automatically set by the rest application to the FIRST listener supplied in the standard `listener` config.

As the `listener` setting can use automatic port allocation, e.g. if set to `http://localhost:0` then it will auto-allocate a port number, we might not know the actual port until the Jetty server is started.
This causes problems where components require the `KsqlConfig` to be injected at creation time. To get around this I've added `KsqlConfigurable`, which components can implement, meaning they get passed their config after creation, and that config includes the required app server setting with the correct port.

The following components currently implement `KsqlConfigurable`:

* `KsqlResource`
* `StreamedQueryResource`
* `StatementExecutor`

Strictly speaking, only the last of these currently _needs_ the app server set, as it is the only one the _currently_ starts persistent queries. But I think it's safer, i.e. less likely to cause bugs in the future, if all three are configured with the app server set.

### Testing done 

Some tests added in this PR. More coming with later point query PRs.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java [hunk 1]
```diff
@@ -115,7 +116,7 @@
 
   public static final String COMMANDS_STREAM_NAME = "KSQL_COMMANDS";
 
-  private final KsqlConfig ksqlConfig;
+  private final KsqlConfig ksqlConfigNoPort;
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: I've renamed this to make it clear it does not contain the port info.  My thinking being that people may think twice before using it blindly.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java [hunk 2]
```diff
@@ -198,6 +195,8 @@ public void setupResources(final Configurable<?> config, final KsqlRestConfig ap
   @Override
   public void start() throws Exception {
     super.start();
+    final KsqlConfig ksqlConfigWithPort = buildConfigWithPort();
+    configurables.forEach(c -> c.configure(ksqlConfigWithPort));
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Once the Jetty server has been started we know the actual ports, so can now build more complete config and configure those components that are lazily configured.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/util/ClusterTerminatorTest.java [hunk 3]
```diff
@@ -109,7 +109,6 @@ public void setup() {
     when(serviceContext.getTopicClient()).thenReturn(kafkaTopicClient);
     when(serviceContext.getSchemaRegistryClient()).thenReturn(schemaRegistryClient);
     clusterTerminator = new ClusterTerminator(
-        ksqlConfig,
```
Comment (big-andy-coates) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/util/ClusterTerminatorTest.java]: removed, as its not used.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java [hunk 4]
```diff
@@ -109,26 +127,38 @@ public KsqlResource(
...
         Objects.requireNonNull(activenessRegistrar, "activenessRegistrar");
+    this.injectorFactory = Objects.requireNonNull(injectorFactory, "injectorFactory");
+    this.topicAccessValidator = Objects
+        .requireNonNull(topicAccessValidator, "topicAccessValidator");
+  }
+
+  @Override
+  public void configure(final KsqlConfig config) {
+    if (!config.getKsqlStreamConfigProps().containsKey(StreamsConfig.APPLICATION_SERVER_CONFIG)) {
+      throw new IllegalArgumentException("Need KS application server set");
```
Comment (agavra) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java]: nit: change this to include the config name

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java [hunk 5]
```diff
@@ -624,4 +625,22 @@ private static void maybeCreateProcessingLogStream(
...
+  /**
+   * Build a complete config with the KS IQ application.server set.
+   *
+   * @return true server config.
+   */
+  private KsqlConfig buildConfigWithPort() {
+    final Map<String, Object> props = ksqlConfigNoPort.originals();
+
+    // Wire up KS IQ endpoint discovery to the FIRST listener:
+    final URL firstListener = getListeners().get(0);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Good idea.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java (hunk 4 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java]: It's not something the client will ever see, or if they did it's nothing they could fix.   If its not set, it's a programming error.

**Pull request merged**